src/optimizer/loadOptimizers.ts

```diff
@@ -0,0 +1,79 @@
+import { OptimizerPlugin } from './OptimizerPlugin';
+
+import fs from 'fs-extra';
+import path from 'path';
+import { uniq } from 'lodash';
+import toposort from 'toposort';
+import { logger } from '../utils/GoFSHLogger';
+
+/**
+ * Dynamically load the optimizer plugins and return them in the order they should be executed.
+ * @param folder - the folder to load optimizers from.  Currently only used in tests.  CLI uses default value.
+ */
+export async function loadOptimizers(
+  folder = path.join(__dirname, 'plugins')
+): Promise<OptimizerPlugin[]> {
+  // make an import-friendly relative path (e.g. \Users\bob\dev\optimizers --> ../../../../dev/optimizers)
+  let relativePath = path.relative(__dirname, folder);
+  if (path.sep === '\\') {
+    relativePath = relativePath.replace(/\\/g, '/');
+  }
+  if (!relativePath.startsWith('.')) {
+    relativePath = `./${relativePath}`;
+  }
+
+  // read the plugins folder, filter to only .ts/.js files, and map to unique module paths (without extensions)
+  const optimizerModules = uniq(
+    fs
+      .readdirSync(folder)
+      .filter(f => /^[^.]+.*\.(t|j)s$/.test(f))
+      .map(f => `${relativePath}/${f.match(/^([^.]+)/)[1]}`)
+  );
+
+  const optimizers = (
+    await Promise.all(
+      // map to a set of promises that each resolve to the default export of the dynamically imported module
+      optimizerModules.map(async m => {
+        return (await import(m))?.default as OptimizerPlugin;
+      })
+    )
+  ).filter(
+    // Remove non-optimizers
+    o =>
+      typeof o?.name === 'string' &&
+      typeof o?.description === 'string' &&
+      typeof o?.optimize === 'function'
+  );
+  logger.debug(`Loaded ${optimizers.length} optimizers from ${path.join(__dirname, 'plugins')}`);
+
+  // Sort them using a topological sort to get them in dependency order
+  // See: https://www.npmjs.com/package/toposort#sorting-dependencies
+  const nodes: string[] = [];
+  const edges: [string, string][] = [];
+  optimizers.forEach(opt => {
+    nodes.push(opt.name);
+    opt.runAfter?.forEach(dependsOn => edges.push([opt.name, dependsOn]));
+    opt.runBefore?.forEach(dependedOnBy => edges.push([dependedOnBy, opt.name]));
+  });
+  let ordered: string[];
+  try {
+    ordered = toposort.array(nodes, edges).reverse();
+  } catch (e) {
+    const nodeMatch = e.message?.match(/"([^"]+)"$/);
+    if (nodeMatch?.length === 2) {
+      logger.error(
+        `Could not determine order of optimizers; cyclic dependency involving ${nodeMatch[1]}. Optimization may be affected.`
+      );
+    } else {
+      logger.error(
+        'Could not determine order of optimizers due to a cyclic dependency. Optimization may be affected.',
+        e
+      );
```

Comment:
Ah.  That's an excellent thing I need to check for (references to non-existent optimizers).  I think I should do that before doing the `toposort` so it can have a more reasonable error message and so I can just remove the bad reference before doing the `toposort` (so the sort has a chance of succeeding).

Reply:
> What does the input look like in a case where there is a cyclic dependency, but the error message does not know any of the involved nodes?

I reviewed the `toposort` source code to see if the error message was reliably consistent and found that there was a case in which a circular dependency error might be returned without referencing a specific node ([source code](https://github.com/marcelklehr/toposort/blob/master/index.js#L38-L46)).  Upon closer inspection, however, I see that would only happen if `JSON.stringify` threw an error -- which seems highly unlikely since our nodes are `string`s.  So perhaps this extra case is not needed.

Reply:
And upon reviewing `toposort` code again, I see it will also throw on an "unknown node" -- which my code will catch and mistakenly report as a circular dependency.  As long as I do a pre-check of nodes and edges, however, that particular error should be unreachable.

Reply:
@mint-thompson -- addressed in c27ec30.